### ISSUE TITLE
Error message `[TS2531] Object is possibly 'null'.` is not always reported by the TypeScript compiler for JS source files

### ISSUE THREAD

**Issue by gladiator001** [Microsoft/TypeScript#27451]
**TypeScript Service Version:** 
3.0.1

**Search Terms:**
TS2531, control flow analysis (CFA)

**Code**

```js
// JS code
(function () {
    "use strict";

    function objectIsPossiblyNullReported() {
        var a = null;
        return a.foo; // Error message `[TS2531] Object is possibly 'null'.` is reported. 👍
    }

    objectIsPossiblyNullReported();

    function objectIsPossiblyNullNotReported(param) {
        var hashTable = null;

        if (param) {
            hashTable = {
                foo: "lorem ipsum"
            };
        }

        return hashTable.foo; // Error message `[TS2531] Object is possibly 'null'.` is NOT reported. 👎 
    }

    objectIsPossiblyNullNotReported(undefined);
})();
```
**Compiler options specified in the `tsconfig.json` file:**

```json
"compilerOptions": {
  "allowJs": true,
  "checkJs": true,
  "alwaysStrict": true,
  "target": "es5"
}
```

**Expected behavior:**
The error message `[TS2531] Object is possibly 'null'.` should be reported for the `objectIsPossiblyNullNotReported` function as `hashTable` can be `null`.

**Actual behavior:**
The error message `[TS2531] Object is possibly 'null'.` is NOT reported for the `objectIsPossiblyNullNotReported` function.

**Playground Link:**
This issue only affects JS code.

**Related Issues:** No.

**Visual Studio Code screenshot:**
<img width="1164" alt="ts2531" src="https://user-images.githubusercontent.com/3323182/46249117-0f569780-c424-11e8-88c0-568dec3d7b21.png">

**Comment by gladiator001** [comment]
@RyanCavanaugh My intention with the`"alwaysStrict": true` compiler option is to:
 > Parse in strict mode and emit "use strict" for each source file.

as documented.
https://www.typescriptlang.org/docs/handbook/compiler-options.html

Enabling `"strict": true` is also enabling `--noImplicitAny`, `--noImplicitThis`, `--alwaysStrict`, `--strictNullChecks`, `--strictFunctionTypes` and `--strictPropertyInitialization`, which is not what I want.

According to the current documentation of `strictNullChecks`:

> In strict null checking mode, **the null and undefined values are not in the domain of every type** and are only assignable to themselves and any (the one exception being that undefined is also assignable to void).

However, in my case, `null` and `undefined` values are in the domain of every type.
IMHO, the error message `[TS2531] Object is possibly 'null'.` should be reported regardless of the `strictNullChecks` compiler option as in the `objectIsPossiblyNullReported` function.

**Comment by RyanCavanaugh** [comment]
> IMHO, the error message [TS2531] Object is possibly 'null'. should be reported regardless of the strictNullChecks compiler option

The *entire point* of `strictNullChecks: false` is to not issue those sorts of errors.
